view/src/extras/address_autocomplete_field.js

```diff
@@ -0,0 +1,178 @@
+import React from "react";
+import TextField from "@material-ui/core/TextField";
+import Autocomplete from "@material-ui/lab/Autocomplete";
+import LocationOnIcon from "@material-ui/icons/LocationOn";
+import Grid from "@material-ui/core/Grid";
+import Typography from "@material-ui/core/Typography";
+import { makeStyles } from "@material-ui/core/styles";
+import parse from "autosuggest-highlight/parse";
+import throttle from "lodash/throttle";
+
+/**
+ * This function creates a script tag with the src and id
+ * set to the value of the parameters. It is the child of
+ * of the position parameter
+ * @param src       The source for the script tag
+ * @param position  The parent object of the script tag
+ * @param id        The id of the script tag
+ */
+function loadScript(src, position, id) {
+  if (!position) {
+    return;
+  }
+
+  const script = document.createElement("script");
+  script.setAttribute("async", "");
+  script.setAttribute("id", id);
+  script.src = src;
+  position.appendChild(script);
+}
+
+const autocompleteService = { current: null };
+
+const useStyles = makeStyles((theme) => ({
+  icon: {
+    color: theme.palette.text.secondary,
+    marginRight: theme.spacing(2),
+  },
+}));
+
+/**
+ * This functional component exports a form input field that autocompletes
+ * user's location queries. It queries the Google place API and uses the
+ * handle location prop passed from the parent to save the desired information
+ * from the resulting locaiton.
+ * @param props Used to pass default location from parent and function
+ *              to handle location change in parent
+ */
+export default function AddressAutocompleteField(props) {
+  const classes = useStyles();
+  const [value, setValue] = React.useState(props.location);
+  const [inputValue, setInputValue] = React.useState("");
+  const [options, setOptions] = React.useState([]);
+  const loaded = React.useRef(false);
+
+  // add the script tag using loadScript
+  if (typeof window !== "undefined" && !loaded.current) {
+    if (!document.querySelector("#google-maps")) {
+      loadScript(
+        "https://maps.googleapis.com/maps/api/js?key=AIzaSyDSPAadPOPcdIXimF1DTzKcHcp7TKa0ht8&libraries=places",
+        document.querySelector("head"),
+        "google-maps"
+      );
+    }
+
+    loaded.current = true;
+  }
+
+  // used to memoize results from API (for optimization)
+  const fetch = React.useMemo(
```

Comment:
I made this comment on some other PR, but I can no longer find it (or your answer if you made one), so...  how does memoizing this actually memoize the result from the server?  It seems to me like it should memoize the creation of the `throttle` function, but it will still _call_ the throttle function each time.  It's a little hard to reason about because `useMemo` is taking a function that returns a function, but unless it somehow indirects through _both_ function calls and memoizes the value it gets from calling the _inner_ function, I'm not sure how this works.

Reply:
Here is my response from last time: 
_
https://reactjs.org/docs/hooks-reference.html#usememo

You're exactly right that this is used for memoizing. The reason a function is passed to useMemo is that this function is run during rendering.
_

To summarize, useMemo is given a function that will run at rendering and the value returned from the function run at rendering is used for memoization (at least this is what is happening to my best understanding).